### PR TITLE
[storage][vineyard] Implement storage lib of vineyard backend

### PR DESCRIPTION
**Pull request by acezen**
## What do these changes do?

implement vineyard backend storage lib.

- [x] vineyard backend
- [x] unit test case

this pr should be merged after #1904 

## Related issue number
#1905

### CODE REVIEW

**Comment by hekaisheng** [comment]
#1904 has been merged, there's some small changes in storage lib API.

**Review by qinxuye — COMMENTED** [review]
LGTM overall, I left some comments.

**Review by qinxuye — COMMENTED** [review]
I am wondering why etcd is required for core testing?

**Comment by qinxuye** [comment]
I think vineyard storage can be tested in vineyard platform testing only.

File: .github/workflows/core-ci.yml [hunk 1]
```diff
@@ -80,6 +80,16 @@ jobs:
         retry ./.github/workflows/download-etcd.sh
         conda list -n test
 
+        # install etcd
```
[the remark this answers is not in the document]
Comment (acezen) [reply .github/workflows/core-ci.yml:83]: I just want to test  vineyard storage lib test on other testing, but it seems break down other tests, I will remove the etcd and move the vineyard storage to  vineyard platform testing.

File: mars/storage/tests/test_libs.py [hunk 2]
```diff
@@ -80,9 +92,10 @@ async def test_base_operations(storage_context):
     info2 = await storage.object_info(put_info2.object_id)
     assert info2.size == put_info2.size
 
-    num = len(await storage.list())
-    assert num == 2
-    await storage.delete(info2.object_id)
+    if not isinstance(storage, VineyardStorage):
```
[the remark this answers is not in the document]
Comment (acezen) [reply mars/storage/tests/test_libs.py:105]: done.

**Review by sighingnow — DISMISSED** [review]
Overall LGTM. Just one concern is should be impose another storage (maybe `VINEYARD`) and use it for vineyard?

Or `MEMORY` is fine enough here?

I would like to know comments  from @qinxuye.

**Review by sighingnow — COMMENTED** [review]
Could we add vineyard (maybe a fixed version) to the requirements.txt of mars?

**Comment by qinxuye** [comment]
> Overall LGTM. Just one concern is should be impose another storage (maybe `VINEYARD`) and use it for vineyard?
> 
> Or `MEMORY` is fine enough here?
> 
> I would like to know comments from @qinxuye.

You mean storage level? I think MEMORY is enough.

**Comment by qinxuye** [comment]
> Could we add vineyard (maybe a fixed version) to the requirements.txt of mars?

As for requirements, it's about necessary libraries, how about add a `pip install mars[vineyard]` support?

**Comment by sighingnow** [comment]
> You mean storage level? I think MEMORY is enough.

Then this pull request looks to me.

> As for requirements, it's about necessary libraries, how about add a `pip install mars[vineyard]` support?

Add a `vvineyard` profile is also ok. I'm wondering if we could put vineyard at the same dependency level with plasma.

**Comment by qinxuye** [comment]
> > You mean storage level? I think MEMORY is enough.
> 
> Then this pull request looks to me.
> 
> > As for requirements, it's about necessary libraries, how about add a `pip install mars[vineyard]` support?
> 
> Add a `vvineyard` profile is also ok. I'm wondering if we could put vineyard at the same dependency level with plasma.

Yeah, but for now, Mars does not rely on plasma either, but in the near future, multiprocessing will be the default option, till then, we can see if vineyard could be the default library.

**Comment by qinxuye** [comment]
@wjsi Could you please help to check why vineyard storage is not covered?

**Comment by sighingnow** [comment]
> @wjsi Could you please help to check why vineyard storage is not covered?

I'm not very sure, I think it is becase there are three `pytest` call and the result for the first one is overrided by the later run. Maybe we need to combine like https://github.com/mars-project/mars/blob/master/.github/workflows/run-tests.sh#L16

**Comment by qinxuye** [comment]
> > @wjsi Could you please help to check why vineyard storage is not covered?
> 
> I'm not very sure, I think it is becase there are three `pytest` call and the result for the first one is overrided by the later run. Maybe we need to combine like https://github.com/mars-project/mars/blob/master/.github/workflows/run-tests.sh#L16

Perhaps, is it possible to combine them into one? @acezen

**Comment by acezen** [comment]
> > > @wjsi Could you please help to check why vineyard storage is not covered?
> > 
> > 
> > I'm not very sure, I think it is becase there are three `pytest` call and the result for the first one is overrided by the later run. Maybe we need to combine like https://github.com/mars-project/mars/blob/master/.github/workflows/run-tests.sh#L16
> 
> Perhaps, is it possible to combine them into one? @acezen

ok

**Comment by qinxuye** [comment]
Hi, @acezen, @wjsi is on it about combing coverage

**Review by qinxuye — APPROVED** [review]
LGTM

**Pull request merged**